packages/spacecat-shared-utils/src/log-wrapper.js

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2024 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+/**
+ * Wraps the provided function to enhance the logging functionality by including the
+ * `jobId` in each log statement.
+ *
+ * @param {function} fn - The original function that will be wrapped.
+ * @returns {function(object, object): Promise<Response>} A new function that, when invoked,
+ *  will automatically include the `jobId` in all log statements within the context.
+ */
+export function logWrapper(fn) {
+  return async (message, context) => {
+    const { log } = context;
+
+    if (log) {
+      if (!message || !message.jobId) {
+        log.debug('Missing jobId, hence it will not be included in log messages.');
+        context.contextualLog = log;
+
+        return fn(message, context);
+      }
+
+      const { jobId } = message;
+
+      // Enhance the log object to include jobId in each log statement
+      context.contextualLog = {
+        info: (...args) => {
+          log.info({ jobId, ...args });
+        },
+        error: (...args) => {
+          log.error({ jobId, ...args });
+        },
+        debug: (...args) => {
+          log.debug({ jobId, ...args });
+        },
+        warn: (...args) => {
+          log.warn({ jobId, ...args });
+        },
```

Comment:
While less frequently used, there are additional log levels. For completeness sake should we provide support for them as well?

Reply:
These are the log levels and methods supported by `console`. Which ones do we want to support? So far, I’ve only seen `info`, `debug`, and `error` being used.

```
console.log()
console.info()
console.warn()
console.error()
console.debug()
console.trace()
console.group()
console.groupEnd()
console.groupCollapsed()
console.table()
console.time()
console.timeEnd()
console.timeLog()
console.assert()
console.clear()
console.count()
console.countReset()
console.dir()
console.dirxml()
console.profile()
console.profileEnd()
console.timeStamp()
```

Reply:
@blefebvre, which log levels should we support? So far, I’ve seen `info`, `debug`, and `error` being used, but it could be extended to include more levels like so:
```js
context.contextualLog = {
  info: (...args) => {
    log.info({ jobId, ...args });
  },
  error: (...args) => {
    log.error({ jobId, ...args });
  },
  debug: (...args) => {
    log.debug({ jobId, ...args });
  },
  warn: (...args) => {
    log.warn({ jobId, ...args });
  },
  trace: (...args) => {
    log.trace({ jobId, ...args });
  },
  log: (...args) => {
    log.log({ jobId, ...args });
  },
  group: (...args) => {
    log.group({ jobId, ...args });
  },
  groupEnd: () => {
    log.groupEnd();
  },
  groupCollapsed: (...args) => {
    log.groupCollapsed({ jobId, ...args });
  },
  assert: (condition, ...args) => {
    log.assert(condition, { jobId, ...args });
  },
  count: (...args) => {
    log.count({ jobId, ...args });
  },
  countReset: (...args) => {
    log.countReset({ jobId, ...args });
  },
  time: (...args) => {
    log.time({ jobId, ...args });
  },
  timeEnd: (...args) => {
    log.timeEnd({ jobId, ...args });
  },
  timeLog: (...args) => {
    log.timeLog({ jobId, ...args });
  },
};
```

Reply:
I’ve added the trace level to the contextual logger. I haven’t seen usage for the other functions so far, but we can update the logger if the need arises later. What do you think?